File: exercises/week02/src/solid/SmokeSensor.java
package solid;

import java.util.Random;

/**
 * A Sensor implementation that detects smoke.
 */
class SmokeSensor implements Sensor {
    private final String location;

    /**
     * A random number generator to simulate a change of detecting smoke.
     */
    private final Random random;
    private int batteryPercentage;

    public SmokeSensor(String location) {
        this.location = location;
        batteryPercentage = 100;
        random = new Random();
    }

    @Override
    public boolean isTriggered() {
        batteryPercentage -= 20;
        return random.nextInt(100) < 10;
    }

    @Override
    public String getLocation() {
        return location;
    }

    @Override
    public String getSensorType() {
        return "Smoke sensor";
    }

    @Override
    public int getBatteryPercentage() {
        return batteryPercentage;
    }
}

File: exercises/week02/src/solid/SmokeSensor.scala
import scala.util.Random

class SmokeSensor extends Sensor {

  var battery = 100.0

  override def isTriggered: Boolean = {
    val random = new Random()
    battery -= 20
    random.nextInt(100) < 10
  }

  override def getLocation: String = "unknown location"

  override def getSensorType: String = "smoke sensor"

  override def getBatteryPercentage: Double = battery
}
